File: BinaryTree/python/binary_tree.py
import typing

E = typing.TypeVar("E")


class Node(typing.Generic[E]):

    def __init__(
            self,
            element: E,
            tree: typing.Optional["BinaryTree"] = None,
            parent: typing.Optional["Node[E]"] = None,
            left: typing.Optional["Node[E]"] = None,
            right: typing.Optional["Node[E]"] = None,
    ) -> None:
        self._element = element
        self._tree = tree
        self._parent = parent
        self._left = left
        self._right = right

    @property
    def element(self) -> E:
        return self._element

    @property
    def tree(self) -> "BinaryTree":
        return self._tree

    @tree.setter
    def tree(self, tree: "BinaryTree") -> None:
        if self._tree:
            raise ValueError("Tree is not empty!")
        self._tree = tree

    @property
    def parent(self) -> typing.Union["Node[E]", None]:
        return self._parent

    @parent.setter
    def parent(self, node: "Node[E]") -> None:
        if self._parent:
            raise ValueError("Parent is not empty!")
        self._parent = node

    @property
    def left(self) -> typing.Union["Node[E]", None]:
        return self._left

    @left.setter
    def left(self, node: "Node[E]") -> None:
        if self._left:
            raise ValueError("Left is not empty!")
        self._left = node

    @property
    def right(self) -> typing.Union["Node[E]", None]:
        return self._right

    @right.setter
    def right(self, node: "Node[E]") -> None:
        if self._right:
            raise ValueError("Right is not empty!")
        if self._left is None:
            raise ValueError("Left is empty!")
        self._right = node

    def children(self) -> list["Node[E]"]:
        output = []
        if self._left:
            output.append(self._left)
            if self._right:
                output.append(self._right)
        return output

    def has_children(self) -> bool:
        return len(self.children()) > 0

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self._element})"


class BinaryTree(typing.Generic[E]):

    def __init__(self, root: typing.Optional[Node[E]] = None) -> None:
        if root is not None:
            root.tree = self
        self._root = root

    @property
    def depth(self) -> int:
        return len(self.get_layers())

    @property
    def height(self) -> int:
        layers = self.get_layers()
        return sum(len(lst) for lst in layers)

    @property
    def size(self) -> int:
        return self.height

    @property
    def empty(self) -> bool:
        return self._root is None

    @property
    def root(self) -> typing.Union[Node[E], None]:
        return self._root

    @root.setter
    def root(self, node: None) -> None:
        if self._root is not None:
            raise ValueError("Tree is not empty!")
        self._root = node

    def _validate(self, node: Node[E]) -> None:
        if node.tree != self:
            raise ValueError("Node does not belong to this tree!")

    def add_left(self, parent: Node[E], element: E) -> Node[E]:
        self._validate(parent)
        if parent.left is not None:
            raise ValueError("Parents left is not empty!")
        node = Node(element, self, parent)
        parent.left = node
        return node

    def add_right(self, parent: Node[E], element: E) -> Node[E]:
        self._validate(parent)
        if parent.right is not None:
            raise ValueError("Parents right is not empty!")
        node = Node(element, self, parent)
        parent.right = node
        return node

    def is_root(self, node: Node[E]) -> bool:
        return (node == self._root) and (self._root is not None)

    @staticmethod
    def is_leaf(node: Node[E]) -> bool:
        return not node.has_children()

    @staticmethod
    def node_height(node: Node[E]) -> int:
        if node is None:
            return 0
        elif not node.has_children():
            return 1
        return 1 + max(BinaryTree.node_height(node.left), BinaryTree.node_height(node.right))

    def get_layers(self) -> list[list[Node[E]]]:
        if self.empty:
            return []
        elif self.is_leaf(self._root):
            return [[self._root]]
        output = [[self._root]]
        children = self._root.children()
        while len(children) > 0:
            output.append(children)
            children_temp = []
            for c in children:
                children_temp.extend(c.children())

            children = children_temp

        return output

    def inorder_traversal(self, node: typing.Optional[Node[E]] = None) -> list[Node[E]]:
        if self.empty:
            return []
        if node is None:
            node = self._root
        output = []
        if node.left:
            output.extend(self.postorder_traversal(node.left))
        output.append(node)
        if node.right:
            output.extend(self.postorder_traversal(node.right))
        return output

    def preorder_traversal(self, node: typing.Optional[Node[E]] = None) -> list[Node[E]]:
        if self.empty:
            return []
        if node is None:
            node = self._root
        output = [node]

        children = node.children()
        for child in children:
            output.extend(self.preorder_traversal(child))

        return output

    def postorder_traversal(self, node: typing.Optional[Node[E]] = None) -> list[Node[E]]:
        if self.empty:
            return []
        if node is None:
            node = self._root
        if self.is_leaf(node):
            return [node]
        output = []
        if node.left:
            output.extend(self.postorder_traversal(node.left))
        if node.right:
            output.extend(self.postorder_traversal(node.right))
        output.append(node)
        return output

    def breath_first_traversal(self) -> list[Node[E]]:
        layers = self.get_layers()
        return [node for lst in layers for node in lst]

    @staticmethod
    def visualize_left_to_right(node: typing.Optional[Node[E]], level: typing.Optional[int]) -> str:

        def _vis(n: Node[E], lvl: int) -> list[str]:
            output = []
            if n.right:
                output.extend(_vis(n.right, lvl + 2))
            output.append(f"""{((" " * 4) * lvl)}-> {n.element}""")
            if n.left:
                output.extend(_vis(n.left, lvl + 2))
            return output

        if node is None:
            return ""
        results = _vis(node, level)
        return "\n".join(results)

    @staticmethod
    def visualize_top_to_bottom(node: typing.Optional[Node[E]], depth: int, start: typing.Optional[int] = 0) -> str:

        def _vis(n: Node[E], d: int, s: int) -> list[str]:
            val = f"{n.element}" if n is not None else " "
            col = d ** 2
            output = [(" " * col) + (" " * s), val]
            output.extend(_vis(n.left, d - 1, col ))

            return output

    def __iter__(self) -> typing.Iterator[Node[E]]:
        return iter(self.breath_first_traversal())

    def __str__(self) -> str:
        return self.visualize_left_to_right(self._root, 0)


if __name__ == "__main__":
    btree: BinaryTree[int] = BinaryTree(Node(0))
    btree.add_left(btree.root, 1)
    btree.add_right(btree.root, 2)

    btree.add_left(btree.root.left, 3)
    btree.add_right(btree.root.left, 4)

    btree.add_left(btree.root.right, 5)
    btree.add_right(btree.root.right, 6)

    btree.add_left(btree.root.left.left, 7)
    print(btree.depth)
    print(btree.size)
    print(btree.inorder_traversal())
    print(btree.preorder_traversal())
    print(btree.postorder_traversal())
    print(btree.breath_first_traversal())
    print()
    print(btree)

File: BinaryTree/dart/bin/binary_tree.dart
import "dart:math" as math;

void main() {
  BinaryTree<int> btree = BinaryTree(Node(0));
  btree.addLeft(btree.root!, 1);
  btree.addRight(btree.root!, 2);

  btree.addLeft(btree.root!.left!, 3);
  btree.addRight(btree.root!.left!, 4);

  btree.addLeft(btree.root!.right!, 5);
  btree.addRight(btree.root!.right!, 6);

  btree.addLeft(btree.root!.left!.left!, 7);
  print(btree.depth);
  print(btree.size);
  print(btree.inorderTraversal());
  print(btree.preorderTraversal());
  print(btree.postorderTraversal());
  print(btree.breathFirstTraversal());
  print(btree);
}

class Node<E> {
  E _element;
  BinaryTree<E>? _tree;
  Node<E>? _parent;
  Node<E>? _left;
  Node<E>? _right;

  Node(this._element, [BinaryTree<E>? tree, Node<E>? parent, Node<E>? left, Node<E>? right]) {
    this._tree = tree;
    this._parent = parent;
    this._left = left;
    this._right = right;
  }

  E get element => _element;

  Node<E>? get right => _right;

  set right(Node<E>? node) {
    this._right = node;
  }

  Node<E>? get left => _left;

  set left(Node<E>? node) {
    this._left = node;
  }

  Node<E>? get parent => _parent;

  set parent(Node<E>? node) {
    this._parent = node;
  }

  BinaryTree<E>? get tree => _tree;

  set tree(BinaryTree<E>? tree) {
    if (_tree != null) {
      throw Exception("Tree is not empty!");
    }
    this._tree = tree;
  }

  List<Node<E>> children() {
    List<Node<E>> output = [];
    if (_left != null) {
      output.add(_left!);
      if (_right != null) {
        output.add(_right!);
      }
    }
    return output;
  }

  bool hasChildren() {
    return children().length > 0;
  }

  @override
  String toString() {
    return "Node($_element)";
  }
}

class BinaryTree<E> extends Iterable<Node<E>> {
  Node<E>? _root;

  BinaryTree([Node<E>? root]) {
    root?.tree = this;
    this._root = root;
  }

  int get depth => getLayers(_root).length;

  int get height {
    List<List<Node<E>>> layers = getLayers(_root);
    int value = 0;
    for (List<Node<E>> list in layers) {
      value += list.length;
    }
    return value;
  }

  int get size => height;

  bool get isEmpty => _root == null;

  Node<E>? get root => _root;

  void _validate(Node<E> node) {
    if (node.tree != this) {
      throw Exception("Node does not belong to this tree!");
    }
  }

  Node<E> addLeft(Node<E> parent, E element) {
    _validate(parent);
    if (parent.left != null) {
      throw Exception("Parent's left is not empty!");
    }

    Node<E> node = Node(element, this, parent);
    parent.left = node;
    return node;
  }

  Node<E> addRight(Node<E> parent, E element) {
    _validate(parent);
    if (parent.right != null) {
      throw Exception("Parent's right is not empty!");
    }

    Node<E> node = Node(element, this, parent);
    parent.right = node;
    return node;
  }

  bool isRoot(Node<E> node) => (node == _root) && (_root != null);

  bool isLeaf(Node<E> node) => !node.hasChildren();

  int nodeHeight(Node<E>? node) {
    if (node == null)
      return 0;
    else if (isLeaf(node)) return 1;
    return 1 + math.max(nodeHeight(node.left), nodeHeight(node.right));
  }

  List<List<Node<E>>> getLayers(Node<E>? node) {
    if (node == null)
      return [];
    else if (isLeaf(node))
      return [
        [node]
      ];

    List<List<Node<E>>> output = [
      [node]
    ];
    List<Node<E>> children = node.children();
    while (children.length > 0) {
      output.add(children);
      List<Node<E>> childrenTemp = [];
      for (Node<E> c in children) {
        childrenTemp.addAll(c.children());
      }

      children = childrenTemp;
    }
    return output;
  }

  List<Node<E>> preorderTraversal([Node<E>? node]) {
    if (isEmpty) return [];

    node = (node ?? _root)!;
    List<Node<E>> output = [node];
    List<Node<E>> children = node.children();
    for (Node<E> child in children) {
      output.addAll(preorderTraversal(child));
    }

    return output;
  }

  List<Node<E>> postorderTraversal([Node<E>? node]) {
    if (isEmpty) return [];
    node = (node ?? _root)!;
    if (isLeaf(node)) return [node];

    List<Node<E>> output = [];
    if (node.left != null) output.addAll(postorderTraversal(node.left));
    if (node.right != null) output.addAll(postorderTraversal(node.right));
    output.add(node);
    return output;
  }

  List<Node<E>> inorderTraversal([Node<E>? node]) {
    if (isEmpty) return [];
    node = (node ?? _root)!;
    List<Node<E>> output = [];
    if (node.left != null) output.addAll(postorderTraversal(node.left));
    output.add(node);
    if (node.right != null) output.addAll(postorderTraversal(node.right));
    return output;
  }

  List<Node<E>> breathFirstTraversal() {
    List<List<Node<E>>> layers = getLayers(_root);
    return [for (List<Node<E>> list in layers) ...list];
  }

  String visualizeLeftToRight(Node<E>? node, int level) {
    if (node == null) return "";

    List<String> _vis(Node<E> n, int lvl) {
      List<String> output = [];
      if (n.right != null) {
        output.addAll(_vis(n.right!, lvl + 2));
      }
      output.add("""${((" " * 4) * lvl)}-> ${n.element}""");
      if (n.left != null) {
        output.addAll(_vis(n.left!, lvl + 2));
      }

      return output;
    }

    List<String> results = _vis(node, level);
    results.insert(0, "\n");
    results.add("\n");
    return results.join("\n");
  }

  @override
  String toString() => visualizeLeftToRight(_root, depth);

  @override
  Iterator<Node<E>> get iterator => breathFirstTraversal().iterator;
}
